openedx/course-discovery
# [ENT-915] Full detail search endpoint

This PR introduces 4 new endpoints in total:

- `/api/v1/search/all/details`
- `/api/v1/search/courses/details`
- `/api/v1/search/course_runs/details`
- `/api/v1/search/programs/details`

These are all searchable, however the current implementation does not add any `search_fields` to actually search in -- that can be added pretty easily later.

This implementation also re-uses existing model serializers (to make it way cheaper to implement dev-time wise) and thus the search is retrieving the underlying object through the SQL DB. This is something to watch out for performance-wise, and eventually a move can be made to using just the search index.

They all return _detailed_ versions of their respective resources, using the most detailed version of the model serializers that were already available.

The `all` endpoint in particular aggregates the resources and presents detailed content type-generic responses, which is the whole point of ENT-915.

*Test instructions*:

1. Make a bunch of different courses, course runs, and programs, any way you like.
1. Run course discovery's reindex management command to make sure this data is available to the API.
1. Ensure that the level of detail retrieved from `/api/v1/search/courses/details` matches `/api/v1/courses/`.
1. Ensure that the level of detail retrieved from `/api/v1/search/course_runs/details` matches `/api/v1/course_runs`.
1. Ensure that the level of detail retrieved from `/api/v1/search/programs/details` matches `/api/v1/programs/{uuid}` for a specific program.
1. Ensure that `/api/v1/search/all/details` retrieves everything you'd get from the 3 other endpoints in aggregated form, with the addition of a `content_type` key which tells what type of content that object is.
